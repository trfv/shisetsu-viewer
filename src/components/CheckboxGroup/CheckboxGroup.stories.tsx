import { ChangeEvent, ComponentProps, useState } from "react";
import { Checkbox } from "../Checkbox";
import { CheckboxGroup } from "./CheckboxGroup";

type Props = ComponentProps<typeof CheckboxGroup>;

export default {
  title: "CheckboxGroup",
  component: CheckboxGroup,
};

export const Basic = ({ label }: Props) => {
  const [values, setValues] = useState<string[]>([]);
  const onChange = (event: ChangeEvent<HTMLInputElement>) => {
    setValues((prev) =>
      event.target.checked
        ? [...prev, event.target.value]
        : prev.filter((v) => v !== event.target.value)
    );
  };
  return (
    <CheckboxGroup label={label} values={values} onChange={onChange}>
      <Checkbox label="A" value="a" />
      <Checkbox label="B" value="b" />
      <Checkbox label="C" value="c" />
      <Checkbox label="D" value="d" />
    </CheckboxGroup>
  );
};

Basic.args = {
  label: "label",
};
