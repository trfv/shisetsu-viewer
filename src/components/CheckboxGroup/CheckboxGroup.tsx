import React, { ChangeEvent, Children, cloneElement, FC, ReactNode } from "react";
import { box, BoxSize } from "../Box";
import { Checkbox } from "../Checkbox";
import { SmallLabel } from "../Label";

type Props = {
  label: string;
  values: string[];
  onChange: (event: ChangeEvent<HTMLInputElement>) => void;
  size?: BoxSize;
  children: ReactNode;
};

const mapChildren = (
  children: ReactNode,
  values: Props["values"],
  onChange: Props["onChange"]
): ReactNode => {
  return Children.map(children, (child: any) => {
    if (child) {
      if (child?.props?.children) {
        return cloneElement(child, {
          children: mapChildren(child.props.children, values, onChange),
        });
      }
      if (child?.type?.displayName === Checkbox.displayName) {
        return cloneElement(child, { checked: values.includes(child.props.value), onChange });
      }
    }
    return null;
  });
};

export const CheckboxGroup: FC<Props> = ({ label, values, onChange, size = "auto", children }) => {
  const Box = box(size);
  return (
    <Box display="flex" flexDirection="column">
      <SmallLabel label={label} />
      <Box component="div" display="flex" marginLeft="-12px">
        {mapChildren(children, values, onChange)}
      </Box>
    </Box>
  );
};
