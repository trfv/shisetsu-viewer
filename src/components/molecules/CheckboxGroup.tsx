import FormControl from "@material-ui/core/FormControl";
import FormGroup from "@material-ui/core/FormGroup";
import React, { ComponentProps, FC } from "react";
import Checkbox from "../atoms/Checkbox";
import FormLabel from "../atoms/FormLabel";

type CheckboxProps = ComponentProps<typeof Checkbox>;

type CheckboxGroupProps = {
  label: string;
  checkboxItems: CheckboxProps[];
};

const CheckboxGroup: FC<CheckboxGroupProps> = ({ label, checkboxItems }) => {
  return (
    <FormControl>
      <FormLabel labelText={label} />
      <FormGroup row>
        {checkboxItems.map((props, index) => (
          <Checkbox key={index} {...props} />
        ))}
      </FormGroup>
    </FormControl>
  );
};

export default CheckboxGroup;
