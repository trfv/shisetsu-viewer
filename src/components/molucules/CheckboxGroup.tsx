import FormControl from "@material-ui/core/FormControl";
import FormGroup from "@material-ui/core/FormGroup";
import React, { FC } from "react";
import Checkbox, { CheckboxProps } from "../atoms/Checkbox";
import FormLabel from "../atoms/FormLabel";

type CheckboxGroupProps = {
  label: string;
  checkboxItems: CheckboxProps[];
};

const CheckboxGroup: FC<CheckboxGroupProps> = ({ label, checkboxItems }: CheckboxGroupProps) => {
  return (
    <FormControl>
      <FormLabel labelText={label} />
      <FormGroup row>
        {checkboxItems.map((props, index) => (
          <Checkbox key={index} {...props} />
        ))}
      </FormGroup>
    </FormControl>
  );
};

export default CheckboxGroup;
