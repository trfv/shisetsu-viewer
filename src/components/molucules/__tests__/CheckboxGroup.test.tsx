import "@testing-library/jest-dom/extend-expect";
import { fireEvent, render } from "@testing-library/react";
import React from "react";
import renderer from "react-test-renderer";
import CheckboxGroup from "../CheckboxGroup";

it("snapshot", () => {
  const props = {
    label: "checkbox",
    checkboxItems: [
      {
        label: "checkbox1",
        checked: false,
        onChange: jest.fn(),
      },
      {
        label: "checkbox2",
        checked: false,
        onChange: jest.fn(),
      },
      {
        label: "checkbox3",
        checked: false,
        onChange: jest.fn(),
      },
    ],
  };
  const component = renderer.create(<CheckboxGroup {...props} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it("interaction", () => {
  const props = {
    label: "checkbox",
    checkboxItems: [
      {
        label: "checkbox1",
        checked: false,
        onChange: jest.fn(),
      },
      {
        label: "checkbox2",
        checked: false,
        onChange: jest.fn(),
      },
      {
        label: "checkbox3",
        checked: false,
        onChange: jest.fn(),
      },
    ],
  };
  const checkbox = render(<CheckboxGroup {...props} />);
  expect(checkbox.queryByLabelText("checkbox1")).toBeTruthy();
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  fireEvent.click(checkbox.queryByLabelText("checkbox1")!);
  expect(props.checkboxItems[0].onChange).toHaveBeenCalled();
  expect(props.checkboxItems[1].onChange).not.toHaveBeenCalled();
  expect(props.checkboxItems[2].onChange).not.toHaveBeenCalled();
});
