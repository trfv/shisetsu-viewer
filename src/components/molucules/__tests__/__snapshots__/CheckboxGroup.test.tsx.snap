// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  className="MuiFormControl-root"
>
  <div
    className="MuiBox-root MuiBox-root-1"
  >
    checkbox
  </div>
  <div
    className="MuiFormGroup-root MuiFormGroup-row"
  >
    <label
      className="MuiFormControlLabel-root"
    >
      <span
        aria-disabled={false}
        className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-2 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label"
        >
          <input
            checked={false}
            className="PrivateSwitchBase-input-5"
            data-indeterminate={false}
            disabled={false}
            onChange={[Function]}
            type="checkbox"
          />
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </span>
      </span>
      <span
        className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
      >
        checkbox1
      </span>
    </label>
    <label
      className="MuiFormControlLabel-root"
    >
      <span
        aria-disabled={false}
        className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-2 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label"
        >
          <input
            checked={false}
            className="PrivateSwitchBase-input-5"
            data-indeterminate={false}
            disabled={false}
            onChange={[Function]}
            type="checkbox"
          />
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </span>
      </span>
      <span
        className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
      >
        checkbox2
      </span>
    </label>
    <label
      className="MuiFormControlLabel-root"
    >
      <span
        aria-disabled={false}
        className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-2 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label"
        >
          <input
            checked={false}
            className="PrivateSwitchBase-input-5"
            data-indeterminate={false}
            disabled={false}
            onChange={[Function]}
            type="checkbox"
          />
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </span>
      </span>
      <span
        className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
      >
        checkbox3
      </span>
    </label>
  </div>
</div>
`;
